File: frontend/src/api/companies.ts
import { keepPreviousData, type QueryKey, useMutation, useQuery, useQueryClient } from '@tanstack/react-query'

import { apiGet, apiRequest } from './client'
import { historyKeys } from './history'
import { refreshAfterWrite } from './refresh'

// Mirrors backend/app/api/routes/companies.py: the lightweight Company editor and its list (Task 07).

export interface TaxonomyRef {
  id: string
  label: string
  active: boolean
}

export interface Establishment {
  id: string
  name: string | null
  siret: string | null
  address_line1: string | null
  address_line2: string | null
  postal_code: string | null
  city: string | null
  country: string | null
  // Free text: siège, agence, entrepôt…
  kind: string | null
  is_primary: boolean
}

export type ActivityStatus = 'active' | 'inactive' | 'unknown'

export interface ProspectSummary {
  id: string
  civility: 'mr' | 'ms' | null
  first_name: string | null
  last_name: string | null
  role_label: string | null
  exact_job_title: string | null
  activity_status: ActivityStatus
  contactability_status: 'contactable' | 'do_not_contact'
}

export interface Company {
  id: string
  display_name: string
  legal_name: string | null
  siren: string | null
  website_url: string | null
  email_domain: string | null
  size_label: string | null
  commercial_segment: TaxonomyRef | null
  activity_categories: TaxonomyRef[]
  project_done_with_circoe: string | null
  project_type: string | null
  circoe_references: string | null
  client_approach: string | null
  // Primary first.
  establishments: Establishment[]
  prospect_count: number
  // The first 100, by last then first name.
  prospects: ProspectSummary[]
  created_at: string
  updated_at: string
}

export type EstablishmentInput = Omit<Establishment, 'id'> & { id: string | null }

// POST creates; PUT replaces every field, the categories and the full establishment list (one left out is removed).
export interface CompanyInput {
  display_name: string
  legal_name: string | null
  siren: string | null
  website_url: string | null
  email_domain: string | null
  size_label: string | null
  commercial_segment_id: string | null
  activity_category_ids: string[]
  project_done_with_circoe: string | null
  project_type: string | null
  circoe_references: string | null
  client_approach: string | null
  establishments: EstablishmentInput[]
}

export interface CompanyListItem {
  id: string
  display_name: string
  legal_name: string | null
  siren: string | null
  email_domain: string | null
  commercial_segment_label: string | null
  // City of the primary establishment.
  city: string | null
  establishment_count: number
  prospect_count: number
  updated_at: string
}

export interface CompanyPage {
  items: CompanyListItem[]
  total: number
}

export type SimilarityReason = 'same_company_name' | 'similar_company_name' | 'same_email_domain'

export interface SimilarCompany {
  id: string
  display_name: string
  legal_name: string | null
  email_domain: string | null
  reasons: SimilarityReason[]
}

export const COMPANY_PAGE_SIZE = 50

export const companyKeys = {
  all: ['companies'] as const,
  list: (search: string, offset: number, limit: number) => ['companies', 'list', search, offset, limit] as const,
  detail: (id: string) => ['companies', 'detail', id] as const,
  similar: (name: string, domain: string, exclude: string | null) => ['companies', 'similar', name, domain, exclude] as const,
}

function companyPath(id: string): `/${string}` {
  return `/companies/${encodeURIComponent(id)}`
}

// `limit` up to 200 (the API's maximum), e.g. for a company picker.
export function useCompanies(search: string, offset = 0, limit = COMPANY_PAGE_SIZE) {
  return useQuery({
    queryKey: companyKeys.list(search, offset, limit),
    queryFn: ({ signal }) => {
      const params = new URLSearchParams({ limit: String(limit), offset: String(offset) })
      if (search) params.set('q', search)
      return apiGet<CompanyPage>(`/companies?${params.toString()}`, signal)
    },
    placeholderData: keepPreviousData,
  })
}

export function useCompany(id: string | null) {
  return useQuery({
    queryKey: companyKeys.detail(id ?? ''),
    queryFn: ({ signal }) => apiGet<Company>(companyPath(id ?? ''), signal),
    enabled: id !== null,
  })
}

// Existing companies the one being typed may duplicate ("does it already exist?"), from its name and e-mail domain.
export function useSimilarCompanies(name: string, emailDomain: string, exclude: string | null) {
  return useQuery({
    queryKey: companyKeys.similar(name, emailDomain, exclude),
    queryFn: ({ signal }) => {
      const params = new URLSearchParams({ name, email_domain: emailDomain })
      if (exclude) params.set('exclude', exclude)
      return apiGet<SimilarCompany[]>(`/companies/similar?${params.toString()}`, signal)
    },
    enabled: name.trim().length >= 3 || emailDomain.trim() !== '',
    placeholderData: keepPreviousData,
  })
}

// Every write goes through the audited API; the list, the detail and its history refresh afterwards. Segment/category
// counts in Paramètres change too, so their caches are refreshed as well.
export function useCompanyMutations() {
  const queryClient = useQueryClient()
  const refresh = (...keys: QueryKey[]) => refreshAfterWrite(queryClient, [companyKeys.all, ['settings'], ...keys])
  const saved = (company: Company) => {
    queryClient.setQueryData(companyKeys.detail(company.id), company)
    void refresh(historyKeys.subject('companies', company.id))
  }
  return {
    create: useMutation({
      mutationFn: (input: CompanyInput) => apiRequest<Company>('POST', '/companies', { body: input }),
      onSuccess: saved,
    }),
    update: useMutation({
      mutationFn: ({ id, input }: { id: string; input: CompanyInput }) =>
        apiRequest<Company>('PUT', companyPath(id), { body: input }),
      onSuccess: saved,
    }),
    remove: useMutation({
      mutationFn: (id: string) => apiRequest<undefined>('DELETE', companyPath(id)),
      onSuccess: (_, id) => {
        queryClient.removeQueries({ queryKey: companyKeys.detail(id) })
        void refresh()
      },
    }),
  }
}

File: backend/app/api/routes/companies.py
"""Companies API (`/api/companies`): the lightweight Company editor and its list (Task 07).

`POST` creates a company with its establishments; `PUT` replaces the whole editable state — every
field, the category list and the **full** establishment list (an establishment left out is
removed), so both lists are required there. Business refusals use the stable codes of
`app.api.errors`: 422 `invalid` with `field` (`siren`, `establishments.1.siret`…) and `reason`
(`format`, `checksum`, `webmail`, `repeated`, `unknown`…), 409 `duplicate` naming the company that
holds a SIREN/SIRET, 409 `in_use` when prospects still reference a company to delete, 404.
"""

import uuid
from datetime import datetime
from typing import Annotated

from fastapi import APIRouter, Query, status
from pydantic import BaseModel, Field, StringConstraints

from app.api.dependencies import CurrentActor, SessionDep
from app.api.errors import business_errors
from app.api.history import HistoryCursor, HistoryLimit, HistoryPageOut
from app.models.enums import ActivityStatus, Civility, ContactabilityStatus
from app.services import companies, history
from app.services.companies import CompanyDetail, CompanyInput, EstablishmentInput
from app.services.imports.models import MatchReason

router = APIRouter(prefix="/companies", tags=["companies"])

# Raw input bounds only; the service applies the precise limits with field-level codes.
Line = Annotated[str, StringConstraints(max_length=1000)]
Paragraph = Annotated[str, StringConstraints(max_length=20_000)]
MAX_ITEMS = 100


class TaxonomyRefOut(BaseModel):
    id: uuid.UUID
    label: str
    active: bool


class EstablishmentIn(BaseModel):
    # Omitted for a new establishment.
    id: uuid.UUID | None = None
    name: Line | None = None
    siret: Line | None = None
    address_line1: Line | None = None
    address_line2: Line | None = None
    postal_code: Line | None = None
    city: Line | None = None
    country: Line | None = None
    kind: Line | None = None
    is_primary: bool = False


class CompanyIn(BaseModel):
    display_name: Line
    legal_name: Line | None = None
    siren: Line | None = None
    website_url: Line | None = None
    email_domain: Line | None = None
    size_label: Line | None = None
    commercial_segment_id: uuid.UUID | None = None
    activity_category_ids: list[uuid.UUID] = Field(default_factory=list, max_length=MAX_ITEMS)
    project_done_with_circoe: Paragraph | None = None
    project_type: Paragraph | None = None
    circoe_references: Paragraph | None = None
    client_approach: Paragraph | None = None
    establishments: list[EstablishmentIn] = Field(default_factory=list, max_length=MAX_ITEMS)


class CompanyReplace(CompanyIn):
    """Both lists are required: a partial body must not wipe the establishments or categories."""

    activity_category_ids: list[uuid.UUID] = Field(max_length=MAX_ITEMS)
    establishments: list[EstablishmentIn] = Field(max_length=MAX_ITEMS)


class EstablishmentOut(BaseModel):
    id: uuid.UUID
    name: str | None
    siret: str | None
    address_line1: str | None
    address_line2: str | None
    postal_code: str | None
    city: str | None
    country: str | None
    kind: str | None
    is_primary: bool


class ProspectSummaryOut(BaseModel):
    id: uuid.UUID
    civility: Civility | None
    first_name: str | None
    last_name: str | None
    role_label: str | None
    exact_job_title: str | None
    activity_status: ActivityStatus
    contactability_status: ContactabilityStatus


class CompanyOut(BaseModel):
    id: uuid.UUID
    display_name: str
    legal_name: str | None
    siren: str | None
    website_url: str | None
    email_domain: str | None
    size_label: str | None
    commercial_segment: TaxonomyRefOut | None
    activity_categories: list[TaxonomyRefOut]
    project_done_with_circoe: str | None
    project_type: str | None
    circoe_references: str | None
    client_approach: str | None
    establishments: list[EstablishmentOut]
    prospect_count: int
    # The first 100 prospects, by last then first name.
    prospects: list[ProspectSummaryOut]
    created_at: datetime
    updated_at: datetime


class CompanyListItemOut(BaseModel):
    id: uuid.UUID
    display_name: str
    legal_name: str | None
    siren: str | None
    email_domain: str | None
    commercial_segment_label: str | None
    city: str | None
    establishment_count: int
    prospect_count: int
    updated_at: datetime


class CompanyPageOut(BaseModel):
    items: list[CompanyListItemOut]
    total: int


class SimilarCompanyOut(BaseModel):
    id: uuid.UUID
    display_name: str
    legal_name: str | None
    email_domain: str | None
    reasons: list[MatchReason]


def company_out(detail: CompanyDetail) -> CompanyOut:
    return CompanyOut.model_validate(detail, from_attributes=True)


def company_input(body: CompanyIn) -> CompanyInput:
    return CompanyInput(
        **body.model_dump(exclude={"establishments"}),
        establishments=[EstablishmentInput(**item.model_dump()) for item in body.establishments],
    )


@router.get("")
def list_companies(
    session: SessionDep,
    q: Annotated[str | None, Query(max_length=200)] = None,
    limit: Annotated[int, Query(ge=1, le=companies.LIST_MAX_LIMIT)] = 50,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> CompanyPageOut:
    """Ordered by name; `q` matches every word of the names, e-mail domain and website (case and
    accents ignored), or digits of a SIREN/SIRET."""
    page = companies.list_companies(session, search=q, limit=limit, offset=offset)
    return CompanyPageOut.model_validate(page, from_attributes=True)


@router.get("/similar")
def similar_companies(
    session: SessionDep,
    name: Annotated[str | None, Query(max_length=1000)] = None,
    email_domain: Annotated[str | None, Query(max_length=1000)] = None,
    exclude: uuid.UUID | None = None,
) -> list[SimilarCompanyOut]:
    """Existing companies the typed one may duplicate (warning before creating it)."""
    found = companies.find_similar(
        session, name=name, email_domain=email_domain, exclude_id=exclude
    )
    return [SimilarCompanyOut.model_validate(item, from_attributes=True) for item in found]


@router.get("/{company_id}")
def get_company(company_id: uuid.UUID, session: SessionDep) -> CompanyOut:
    with business_errors():
        return company_out(companies.get_company(session, company_id))


@router.get("/{company_id}/history")
def company_history(
    company_id: uuid.UUID,
    session: SessionDep,
    limit: HistoryLimit = 10,
    before: HistoryCursor = None,
) -> HistoryPageOut:
    """Readable history, newest first: the company's fields, categories and establishments."""
    page = history.history_page(session, "company", company_id, limit=limit, before=before)
    return HistoryPageOut.model_validate(page, from_attributes=True)


@router.post("", status_code=status.HTTP_201_CREATED)
def create_company(body: CompanyIn, session: SessionDep, actor: CurrentActor) -> CompanyOut:
    with business_errors():
        return company_out(companies.create_company(session, actor, company_input(body)))


@router.put("/{company_id}")
def update_company(
    company_id: uuid.UUID, body: CompanyReplace, session: SessionDep, actor: CurrentActor
) -> CompanyOut:
    with business_errors():
        detail = companies.update_company(session, actor, company_id, company_input(body))
    return company_out(detail)


@router.delete("/{company_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_company(company_id: uuid.UUID, session: SessionDep, actor: CurrentActor) -> None:
    with business_errors():
        companies.delete_company(session, actor, company_id)
